New use case: Onboarding of CNFs to the platform (#175)
cncf/cnf-wg: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md

In CNF case the assumption is that the platform is already there, that it is in its core Kubernetes based without special forks as well as that different components needed by CNF could be installed via Kubernetes API.

This creates the situation in which CNF and CSP platform, both of which have their requirements and pre-conditions, meet each other as such for the first time. The process of getting the CNF deployed on the platform in aforementioned scenario is colloquially called "onboarding". In this process CNF vendor, CNF DevOps and CNPD work together to reach common gloal of having production grade deployment of CNF supported with appropriate SLA which factors-in the fact that CNF runs on CSP platform.  

## Expected behaviour

> CNF vendor expects that CSP platform and environment (e.g. Network) are fulfiling certain requirements and preconditions for CNF to be able to achieve production grade deployment. These requirements are different from CNF to CNF.

Author: Anuket to my knowledge is perscriptive when it comes to selection of components and their versions. It does not provide generalized requirements. However, number of requirements could be derived from it.

> CNF vendor expects that CSP platform and environment (e.g. Network) are fulfiling certain requirements and preconditions for CNF to be able to achieve production grade deployment. These requirements are different from CNF to CNF.

Reviewer: ```suggestion
CNF vendor expects that CSP platform and environment (e.g. Network) are fulfilling certain requirements and preconditions for CNF to be able to achieve production grade deployment. These requirements are different from CNF to CNF.
```

CNPD expects that all requirements are well documented with and provided in such form before onboarding starts. The docummentation shall enable CNPD team to verify readiness and prepare the platform (e.g. configure necessary kernel modules, SR-IOV interfaces and similar). In practically all deployments that are meant for production environment the deployment artifacts are stored in private on-prem artifacts and registries. Therefore CNPD expects that CNF vendor provides its artefacts in a form that is typical for cloud native applications (Helm charts, containers, yaml manifests) which could be easily replicated to CSP's on-prem environment (e.g. CNF vendor's registry accessible via Internet from CSPs caching registry).

> CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continous deployment realisations with that particular or similar CNFs as well as comprihensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.

Reviewer 2: ```suggestion
CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continuous deployment realisations with that particular or similar CNFs as well as comprehensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
```

> CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continous deployment realisations with that particular or similar CNFs as well as comprihensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.

Reviewer 3: ```suggestion
CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continous deployment realisations with that particular or similar CNFs as well as comprihensive deployment troubleshooting guide. As already mentioned in [UC0001](0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md), CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
```

Author: for some reason I can not commit your suggestion @electrocucaracha , but it is fine by me. Can someone e.g. @taylor committ it?

> CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continuous deployment realisations with that particular or similar CNFs as well as comprehensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.

Reviewer 2: ```suggestion
CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continuous deployment realisations with that particular or similar CNFs as well as comprehensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continuous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
```

> The starting point of onboarding shall be extension of platfrom with any requiremetns and dependencies that are coming from CNF. When all conditions on the platform level are met, the deployment of CNF shall proceed.

Reviewer 2: ```suggestion
The starting point of onboarding shall be extension of platform with any requiremetns and dependencies that are coming from CNF. When all conditions on the platform level are met, the deployment of CNF shall proceed.
```

> The starting point of onboarding shall be extension of platform with any requiremetns and dependencies that are coming from CNF. When all conditions on the platform level are met, the deployment of CNF shall proceed.

Reviewer: ```suggestion
The starting point of onboarding shall be extension of platform with any requirements and dependencies that are coming from CNF. When all conditions on the platform level are met, the deployment of CNF shall proceed.
```

## Challenges and limitations with Kube-native approach

> Most of the CNFs today, especially complex ones, are packaged in such way that only trained professional service teams of the CNF vendor can reasonably deploy them. The documentation and manuals that come with those CNFs are giving same impression. The onboarding today usually requires intensive interation between CNF vendor, CNF DevOps and CNPD teams and lots of trial/error tuning and manual adaptations on both platform and CNF side.

Reviewer 4: Shouldn't be the DevOps team a cross organisational team?

Author: Hi Gergely,
When CNF DevOps and Cloud-native Platform DevOps work together it is cross-organisational. It is also advisable that CNF DevOps team has also members who possess CKA-like knowledge and who can deal with deployments, but it depends on the size of the organisation.

Does that make sense?
-Vuk

The CNFs are also regularly packaged to be suitable for opinionated installers that are typically part of vendor's own platform offering or particular platform blueprint. The onboarding on CSP's own platform in such cases frequently involves sort of reverse engineering of CNF vendor's installer to come to basic cloud native artefacts and derive requirements from installer's code with the help of vendor's engineers.

> Finally, the CNFs are frequently delivered as Helm charts. However, these Helm charts often do not cover all dependencies that enables reconciling the deployments in propper way. They also install CRDs which are not managed well by Helm.

Reviewer 2: ```suggestion
Finally, the CNFs are frequently delivered as Helm charts. However, these Helm charts often do not cover all dependencies that enables reconciling the deployments in proper way. They also install CRDs which are not managed well by Helm.
```

This situation is in most cases not blocking the onboarding, but is making it laborious and not repeatable.

### Established practices to overcome challenges and limitations

> The practices already exist outside of CNF world, but are not very well followed by CNF vendors. Some of the practices are here:

Reviewer 2: ```suggestion
The practices already exist outside of CNF world, but are not very well followed by CNF vendors. Some of the practices are here:

```

Reviewer 2: for linting

> * https://helm.sh/docs/chart_best_practices/
> * https://www.gitops.tech/
> * https://learnk8s.io/production-best-practices

Reviewer 3: ```suggestion
- [The Chart Best Practices Guide](https://helm.sh/docs/chart_best_practices/)
- [GitOps](https://www.gitops.tech/)
- [Kubernetes production best practices](https://learnk8s.io/production-best-practices)
```

## Expected behaviour

CNF vendor expects that CSP platform and environment (e.g. Network) are fulfilling certain requirements and preconditions for CNF to be able to achieve production grade deployment. These requirements are different from CNF to CNF.

CNPD expects that all requirements are well documented with and provided in such form before onboarding starts. The docummentation shall enable CNPD team to verify readiness and prepare the platform (e.g. configure necessary kernel modules, SR-IOV interfaces and similar). In practically all deployments that are meant for production environment the deployment artifacts are stored in private on-prem artifacts and registries. Therefore CNPD expects that CNF vendor provides its artefacts in a form that is typical for cloud native applications (Helm charts, containers, yaml manifests) which could be easily replicated to CSP's on-prem environment (e.g. CNF vendor's registry accessible via Internet from CSPs caching registry).

> CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continuous deployment realisations with that particular or similar CNFs as well as comprehensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continuous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.

Reviewer 5: Wrong markdown for UC001 link (first brackets, then parentheses)

Finally, the CNFs are frequently delivered as Helm charts. However, these Helm charts often do not cover all dependencies that enables reconciling the deployments in proper way. They also install CRDs which are not managed well by Helm.

This situation is in most cases not blocking the onboarding, but is making it laborious and not repeatable.

### Established practices to overcome challenges and limitations

The practices already exist outside of CNF world, but are not very well followed by CNF vendors. Some of the practices are here:

- [The Chart Best Practices Guide](https://helm.sh/docs/chart_best_practices/)
- [GitOps](https://www.gitops.tech/)
- [Kubernetes production best practices](https://learnk8s.io/production-best-practices)

### What needs to be done differently in order to overcome challenges and limitations

> CNF vendors need to start preparing CNFs for shipping as regular cloud software that is well docummented and that can be deployed and configured by CNF DevOps team. Following established best practices for cloud native software distribution and deployment would improve the situation significantly.

Reviewer 5: Spelling mistake: "docummented"
